bittide-instances/exe/post-vex-riscv-tcp-test/Main.hs

```diff
@@ -0,0 +1,212 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE NumericUnderscores #-}
+
+import Prelude
+
+import Paths_bittide_instances
+
+import Control.Monad (unless)
+import Control.Monad.Extra (forM_)
+import Data.List.Extra
+import Data.Maybe
+import System.Environment (withArgs)
+import System.IO
+import System.IO.Temp
+import System.Process
+import Test.Tasty.HUnit
+import Test.Tasty.TH
+
+data Error = Ok | Error String
+data Filter = Continue | Stop Error
+
+getOpenOcdStartPath :: IO FilePath
+getOpenOcdStartPath = getDataFileName "data/openocd/start.sh"
+
+getGdbProgPath :: IO FilePath
+getGdbProgPath = getDataFileName "data/gdb/smoltcp-hitl-prog"
+
+getTcpSprayPath :: IO FilePath
+getTcpSprayPath = getDataFileName "data/tcpspray/start.sh"
+
+getPicocomStartPath :: IO FilePath
+getPicocomStartPath = getDataFileName "data/picocom/start.sh"
+
+{- | XXX: Currently hardcoded to a very specific position. Maybe we could probe
+       using JTAG to see what device we're connected to?
+-}
+getUartDev :: IO String
+getUartDev = pure "/dev/serial/by-path/pci-0000:00:14.0-usb-0:5.1:1.1-port0"
+
+{- | Copy the GDB program obtained from 'getGdbProgPath' to a temporary file,
+prepend each non-comment, non-empty line with 'echo > {line}\n'. This effectively
+emulates Bash's 'set -x' for the GDB program. This can in turn be used to
+wait for specific commands to be executed, or simply for debugging.
+-}
+withAnnotatedGdbProgPath :: (String -> IO ()) -> IO ()
+withAnnotatedGdbProgPath action = do
```

Comment:
A bunch of this is copy+pasted right? Could you move it into library code instead and import it?